src/app/functions/server/dashboard.ts

```diff
@@ -0,0 +1,153 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { BreachDataTypes } from "../../../utils/breachResolution";
+
+export interface DashboardSummary {
+  dataBreachTotalNum: number;
+  dataBrokerTotalNum: number;
+  totalExposures: number;
+  allExposures: {
+    // shared
+    emailAddresses: number;
+    phoneNumbers: number;
+
+    // data brokers
+    addresses: number;
+    familyMembers: number;
+    fullNames: number;
+
+    // data breaches
+    socialSecurityNumbers: number;
+    ipAddresses: number;
+    passwords: number;
+    creditCardNumbers: number;
+    pinNumbers: number;
+    securityQuestions: number;
+  };
+  sanitizedExposures: Array<Record<string, number>>;
+}
+
+export function dashboardSummary(
+  scannedResults,
+  { breachesData }
+): DashboardSummary {
+  const summary: DashboardSummary = {
+    dataBreachTotalNum: 0,
+    dataBrokerTotalNum: 0,
+    totalExposures: 0,
+    allExposures: {
+      emailAddresses: 0,
+      phoneNumbers: 0,
+      addresses: 0,
+      familyMembers: 0,
+      fullNames: 0,
+
+      // data breaches
+      socialSecurityNumbers: 0,
+      ipAddresses: 0,
+      passwords: 0,
+      creditCardNumbers: 0,
+      pinNumbers: 0,
+      securityQuestions: 0,
+    },
+    sanitizedExposures: [],
+  };
+
+  // calculate broker summary from scanned results
+  if (scannedResults) {
+    scannedResults.forEach((r) => {
+      // count email
+      summary.totalExposures += r.emails.length;
+      summary.allExposures.emailAddresses += r.emails.length;
+
+      // count phones
+      summary.totalExposures += r.phones.length;
+      summary.allExposures.phoneNumbers += r.phones.length;
+
+      // count physical addresses
+      summary.totalExposures += r.addresses.length;
+      summary.allExposures.addresses += r.addresses.length;
+
+      // count relatives
+      summary.totalExposures += r.relatives.length;
+      summary.allExposures.familyMembers += r.relatives.length;
+
+      // count full name
+      summary.totalExposures++;
+      summary.allExposures.fullNames++;
+    });
+  }
+
+  // calculate breaches summary from breaches data
+  if (breachesData.verifiedEmails) {
+    for (const emailBreaches of breachesData.verifiedEmails) {
+      const breaches = emailBreaches.breaches;
+      breaches.forEach((b) => {
+        const dataClasses = b.DataClasses;
```

Comment:
Pretty sure `DataClasses` from HIBP is always defined and non-empty.
Not sure if it's easier to default `dataClasses` here to an empty array so we don't need to use optional chaining `dataClasses?.` 6&times; below.

```js
const dataClasses = b.DataClasses ?? [];
```